tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc

```diff
@@ -92,40 +83,33 @@ Value InsertDynamicAllocAndDealloc(Location loc, Value result,
   return alloc;
 }
 
-Value InsertAllocAndDealloc(Location loc, Value result,
-                            ConversionPatternRewriter* rewriter) {
+Value InsertAlloc(Location loc, Value result,
+                  xla::BufferAssignmentPlacer* bufferAssignment,
+                  ConversionPatternRewriter* rewriter) {
   auto result_type = result.getType().dyn_cast<ShapedType>();
   if (!result_type || !result_type.hasStaticShape()) {
     result.getDefiningOp()->emitOpError()
         << "tensor to buffer conversion expects statically shaped results";
   }
   auto memref_type =
       MemRefType::get(result_type.getShape(), result_type.getElementType());
-
-  Operation* op = result.getDefiningOp();
-  auto block = op->getBlock();
-
-  OpBuilder allocBuilder(op);
-  allocBuilder.setInsertionPointToStart(block);  // Inserting at the beginning
+  auto allocPosition = bufferAssignment->computeAllocPosition(result);
```

Comment:
Looking at this I realized that this is used by rewrite patterns, so it should not use a builder of its own. Instead, use the `rewriter`. It implements the same interface, so you can also call `restoreInsertionPoint` on it.

Reply:
You should also use an `InsertionGuard` to make sure that your changing of the position does not leak into the calling context. Something like 'OpBuilder::InsertionGuard(rewriter)` at the top should do.

Reply:
Done.